proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+
+### Resources
+
+Further details about the context and goal of the proposed changes can be found in [the talk I gave at HIW 2024](https://www.youtube.com/live/uMurx1a6Zck?si=kNR2Qsj-6hVGdlP2&t=10579) and the related [JFLA 2024 article](https://inria.hal.science/hal-04406360/document).
+
+I built a proof-of-concept library that leverages `compactAddHollow#` to make a destination-passing style interface for compact regions [here](https://github.com/tweag/linear-base/pull/450). It is part of [`linear-base`](https://github.com/tweag/linear-base) as linear types are a key part of the safety of the interface. More formal considerations about the safety of the API should follow soon.
+
+The GHC branch that implements the changes described in the proposal is available [here](https://gitlab.haskell.org/tweag/ghc/-/tree/tbagrel1/dps-compact-regions-prims?ref_type=heads) ([diff](https://gitlab.haskell.org/tweag/ghc/-/compare/master...tbagrel1%2Fdps-compact-regions-prims?from_project_id=1)).
+
+## Proposed Change Specification
+
+1. [Allocate hollow data constructor in region: RTS primop `compactAddHollow#`](#allocate-hollow-data-constructor-in-region-rts-primop-compactaddhollow)
+2. [Retrieving info table pointer of a constructor: primtype `InfoTablePtrOf#` and compile-time primop `reifyInfoTablePtr#`](#retrieving-info-table-pointer-of-a-constructor-primtype-infotableptrof-and-compile-time-primop-reifyinfotableptr)
+3. [Obtain data constructor name as a `Symbol`: `LiftedCtorToSymbol` built-in type family](#obtain-data-constructor-name-as-a-symbol-liftedctortosymbol-built-in-type-family)
+
+### Allocate hollow data constructor in region: RTS primop `compactAddHollow#`
+
+The new RTS primitive `compactAddHollow#` allocates space for a new constructor heap object in the specified compact region, set its header with the supplied info table pointer, but doesn't set any of its fields (so the constructor is said _hollow_). The user is responsible for setting the fields of the constructor heap object before reading it. The primitive returns the newly allocated constructor heap object, but also the size of the header of the object, and the size of the words on the platform so that the user can later set the fields of the returned object.
+
+```haskell
+compactAddHollow#
+  :: Compact#         -- ^ the compact region where the constructor heap object will be added
+  -> Addr#            -- ^ the info table pointer that will be set as header for the new constructor heap object
+  -> State# RealWorld -- ^ state token
+  -> (# State# RealWorld
+      , a             -- ^ the allocated constructor heap object
+      , Word#         -- ^ size of the header of the allocated heap object, in bytes
+      , Word#         -- ^ size of the words, in bytes, on the platform
```

Comment:
An alternative to discuss is to make this word size a primitive, so that you don't have to return it.